framework/test/unit/engine/generator/generator.spec.ts

```diff
@@ -808,5 +809,49 @@ describe('generator', () => {
 				expect(generator['_consensus'].certifySingleCommit).not.toHaveBeenCalled();
 			});
 		});
+
+		describe('when previous finalized height change and maxRemovalHeight is non zero', () => {
+			beforeEach(async () => {
+				await generator.init({
+					blockchainDB,
+					generatorDB,
+					logger,
+					genesisHeight: 0,
+				});
+				await generator.start();
+				jest.spyOn(generator, '_handleFinalizedHeightChanged' as never);
+				jest.spyOn(generator, '_certifySingleCommitForChangedHeight' as never);
+			});
+
+			afterEach(async () => {
+				await generator.stop();
+			});
+
+			it('should not call certifySingleCommit when getMaxRemovalHeight is higher than next finalized height', async () => {
+				jest.spyOn(consensus, 'getMaxRemovalHeight').mockResolvedValue(30000);
+				generator['_consensus'].events.emit(CONSENSUS_EVENT_FINALIZED_HEIGHT_CHANGED, {
+					from: 0,
+					to: 25520,
+				});
+				await new Promise(process.nextTick);
+
+				expect(generator['_handleFinalizedHeightChanged']).toHaveBeenCalledWith(30000, 25520);
+				expect(generator['_certifySingleCommitForChangedHeight']).not.toHaveBeenCalled();
+			});
+
+			it('should call certifySingleCommit when getMaxRemovalHeight is lower than next finalized height', async () => {
+				jest.spyOn(consensus, 'getMaxRemovalHeight').mockResolvedValue(30000);
+				generator['_consensus'].events.emit(CONSENSUS_EVENT_FINALIZED_HEIGHT_CHANGED, {
+					from: 30001,
+					to: 30003,
+				});
+				await new Promise(process.nextTick);
+
+				expect(generator['_handleFinalizedHeightChanged']).toHaveBeenCalledWith(30001, 30003);
+				expect(generator['_certifySingleCommitForChangedHeight']).toHaveBeenCalledTimes(
```

Comment:
and shouldn't we ensure that `_certifySingleCommit` was called with the value 30003?